.ci/magician/cmd/generate_downstream.go

```diff
@@ -0,0 +1,338 @@
+	if _, err := rnr.Run("cp", []string{"-rp", mmLocalPath, mmCopyPath}, nil); err != nil {
+		fmt.Println("Error copying magic modules: ", err)
+		os.Exit(1)
+	}
+
+	if err := rnr.PushDir(mmCopyPath); err != nil {
+		fmt.Println("Error changing directory to copied magic modules: ", err)
+		os.Exit(1)
+	}
+
+	if err := setConfig(rnr); err != nil {
+		fmt.Println("Error setting config: ", err)
+		os.Exit(1)
+	}
+
+	mmRepo := &source.Repo{
+		Name: "magic-modules",
+		Path: mmCopyPath,
+	}
+	commitMessage, err := branchAndCommitMessage(mmRepo, scratchRepo, command, ref, rnr, ctlr)
+	if err != nil {
+		fmt.Println("Error getting branch and commit message: ", err)
+		os.Exit(1)
+	}
+
+	if err := runMake(upstreamRepo, command, rnr); err != nil {
+		fmt.Println("Error running make: ", err)
+		os.Exit(1)
+	}
+
+	if err := pushCommit(upstreamRepo, scratchRepo, baseBranch, command, commitMessage, ref, gh, rnr, ctlr); err != nil {
+		fmt.Println("Error pushing commit: ", err)
+		os.Exit(1)
+	}
+}
+
+func cloneRepo(baseBranch, repo, version string, ctlr *source.Controller) (*source.Repo, *source.Repo, error) {
+	upstreamRepo := &source.Repo{
+		Title:  repo,
+		Branch: baseBranch,
+	}
+	switch repo {
+	case "terraform":
+		if version == "ga" {
+			upstreamRepo.Name = "terraform-provider-google"
+			upstreamRepo.Version = provider.GA
+		} else if version == "beta" {
+			upstreamRepo.Name = "terraform-provider-google-beta"
+			upstreamRepo.Version = provider.Beta
+		} else {
+			return nil, nil, fmt.Errorf("unrecognized version %s", version)
+		}
+		upstreamRepo.Owner = "hashicorp"
+	case "terraform-google-conversion":
+		upstreamRepo.Name = "terraform-google-conversion"
+		upstreamRepo.Owner = "GoogleCloudPlatform"
+	case "tf-oics":
+		if upstreamRepo.Branch == "main" {
+			upstreamRepo.Branch = "master"
+		}
+		upstreamRepo.Name = "docs-examples"
+		upstreamRepo.Owner = "terraform-google-modules"
+	case "tf-cloud-docs":
+		fmt.Println(repo, " is no longer available.")
+		return nil, nil, nil
+	default:
+		return nil, nil, fmt.Errorf("unrecognized repo %s", repo)
+	}
+	ctlr.SetPath(upstreamRepo)
+	if err := ctlr.Clone(upstreamRepo); err != nil {
+		return nil, nil, err
+	}
+	scratchRepo := &source.Repo{
+		Name:    upstreamRepo.Name,
+		Branch:  baseBranch,
+		Owner:   "modular-magician",
+		Path:    upstreamRepo.Path,
+		Version: upstreamRepo.Version,
+	}
+	return upstreamRepo, scratchRepo, nil
+}
+
+func setConfig(rnr *exec.Runner) error {
+	if _, err := rnr.Run("git", []string{"config", "--local", "user.name", "Modular Magician"}, nil); err != nil {
+		return err
+	}
+	if _, err := rnr.Run("git", []string{"config", "--local", "user.email", "magic-modules@google.com"}, nil); err != nil {
+		return err
+	}
+	return nil
+}
+
+func branchAndCommitMessage(mmRepo, scratchRepo *source.Repo, command, ref string, rnr *exec.Runner, ctlr *source.Controller) (string, error) {
```

Comment:
Its a bit confusing to create the scratch repo then mutate it in a sequential function. Could we merge this into the clone REPO function or create a scratch repo afterwards from the original

Reply:
Done